File: sprint6/kotlin/lesson11code2.kt
// Считываются размеры поля и сама карта.
val n: Int = ...         // Число строк и столбцов в карте.
val m: Int = ...         
val field: Array<IntArray> = ...        // Считываем саму карту.
val start_point: Pair<Int, Int> = ...  // Найдём стартовую клетку, в которой стоит звёздочка.

fun escape_from_maze(n: Int, m: Int, field: Array<IntArray>, start_point: Pair<Int, Int>): Pair<Int, Pair<Int, Int>> {
    // Предок вершины (x, y) -- это предыдущая вершина на кратчайшем пути 
    // от стартовой точки до текущей (x, y). 
    // Изначально заполняем любыми невалидными значениями, например, null или -1.
    val prev = Array(n) { IntArray(m) { -1 } }

    // Заполняем массив расстояний нейтральными по отношению к минимуму значениями.
    // Значение x нейтральное, если min(x, y) = y для любого допустимого y.
    val INF = n * m + 10
    val dist = Array(n) { IntArray(m) { INF } }

    // Начинаем обход в ширину (BFS).
    // Заведём очередь, поддерживающую операции push и pop.
    val q = LinkedList<Pair<Int, Int>>()

    q.add(start_point)
    while (q.isNotEmpty()) {
        val current_point = q.pop()
        for (move in movements) {
            val new_point = add(current_point, move)
            if (is_free(new_point) && dist[new_point.first][new_point.second] == INF) {
                q.add(new_point)
                prev[new_point.first][new_point.second] = current_point
                dist[new_point.first][new_point.second] = dist[current_point.first][current_point.second] + 1
                if (is_exit(new_point)) {
                    return Pair(dist[new_point.first][new_point.second], new_point)
                }
            }
        }
    }

    return Pair(INF, Pair(-1, -1))
}

File: sprint6/By lessons/lesson11code2/lesson11code2.swift
import Foundation

// Определяем структуру для точки с координатами (x, y)
struct Point {
    var x: Int
    var y: Int
}

// Задаем возможные направления движения
let movements: [Point] = [Point(x: -1, y: 0), Point(x: 1, y: 0), Point(x: 0, y: -1), Point(x: 0, y: 1)]

func escapeFromMaze(n: Int, m: Int, field: [[Int]], startPoint: Point) -> (Int, Point) {
    // Предок вершины (x, y) -- это предыдущая вершина на кратчайшем пути 
    // от стартовой точки до текущей (x, y). 
    // Изначально заполняем любыми невалидными значениями, например, -1.
    var prev = [[Int]](repeating: [Int](repeating: -1, count: m), count: n)

    // Заполняем массив расстояний нейтральными по отношению к минимуму значениями.
    // Значение x нейтральное, если min(x, y) = y для любого допустимого y.
    let INF = n * m + 10
    var dist = [[Int]](repeating: [Int](repeating: INF, count: m), count: n)

    // Начинаем обход в ширину (BFS).
    // Заводим очередь для хранения точек.
    var q = Queue<Point>()

    q.enqueue(startPoint)
    while !q.isEmpty {
        let currentPoint = q.dequeue()!
        for move in movements {
            let newPoint = Point(x: currentPoint.x + move.x, y: currentPoint.y + move.y)
            if isFree(newPoint) && dist[newPoint.x][newPoint.y] == INF {
                q.enqueue(newPoint)
                prev[newPoint.x][newPoint.y] = currentPoint.x * m + currentPoint.y
                dist[newPoint.x][newPoint.y] = dist[currentPoint.x][currentPoint.y] + 1
                if isExit(newPoint) {
                    return (dist[newPoint.x][newPoint.y], newPoint)
                }
            }
        }
    }

    return (INF, Point(x: -1, y: -1))
}

// Добавьте функции isFree() и isExit() для проверки свободных клеток и выхода

// Пример использования
let n: Int = ... // Число строк и столбцов в карте.
let m: Int = ...
let field: [[Int]] = ... // Считываем саму карту.
let startPoint: Point = ... // Найдём стартовую клетку, в которой стоит звёздочка.

let result = escapeFromMaze(n: n, m: m, field: field, startPoint: startPoint)
print("Дистанция до выхода: \(result.0)")
print("Координаты выхода: (\(result.1.x), \(result.1.y))")
